tfjs-layers/src/layers/nlp/tokenizers_utils.ts

```diff
@@ -141,3 +141,162 @@ export class BytePairTokenizerCache {
+    }
+    return strs.map(str => str.split(delimRegexPattern).filter(s => s));
+  }
+
+  let regexPattern = delimRegexPattern as unknown as RegExp;
+  if (!regexPattern.flags.includes('g')) {
+    regexPattern = new RegExp(regexPattern.source, regexPattern.flags + 'g');
+  }
+
+  return strs.map(str => {
+    const matches = str.matchAll(regexPattern);
+
+    const splitString = [];
+    let currIdx = 0;
+    for (const match of matches) {
+      splitString.push(str.slice(currIdx, match.index));
+      if (keepDelimRegexPattern) {
+        splitString.push(
+          str.slice(match.index, match.index! + match[0].length));
+      }
+      currIdx = match.index! + match[0].length;
+    }
+    if (currIdx !== str.length) {
+      splitString.push(str.slice(currIdx, str.length));
+    }
+    return splitString.filter(s => s);
+  });
+}
+
+export async function tensorToStringArr(input: Tensor): Promise<string[]> {
+  return await input.data() as unknown as string[];
+}
+
+export async function tensorArrToString2DArr(
+  inputs: Tensor[]): Promise<string[][]> {
+  return Promise.all(inputs.map(
+    async input => await input.data() as unknown as string[]));
+}
+
+export async function splitStringsForBpe(
+  inputs: Tensor, unsplittableTokens?: string[]): Promise<Tensor[]> {
+
+  // We need to recreate the exact behavior of token presplitting in the
+  // original gpt2 implementation which uses a lookahead. We are using an
+  // alternative by inserting a special token "६" before leading space of
+  // non-space characters and after the trailing space, e.g.,
+  // " tf" will be "६ tf".
+  const pattern1 = new RegExp(`( )([^\s${SPECIAL_WHITESPACES}])`);
+  const pattern2 = new RegExp(`(\s${SPECIAL_WHITESPACES})\$`);
+
+  const inputsStr = (await inputs.data() as unknown as string[]).map(str =>
+    str.replace(pattern1, `६$1$2`).replace(pattern2, `$1६`)
+  );
+
+  let alts: string[];
+  let rawTokens: string[][];
+
+  if (unsplittableTokens && unsplittableTokens.length > 0) {
+    alts = createAltsForUnsplittableTokens(unsplittableTokens);
+    unsplittableTokens.forEach((token, idx) => {
+      const alt = alts[idx];
+      const escapedToken = token.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
+
+      rawTokens = regexSplit(rawTokens !== undefined ?
+        rawTokens : inputsStr, escapedToken, escapedToken);
+      rawTokens = rawTokens.map(
+        arr => arr.map(t => t.replace(escapedToken, alt)));
+    });
+  }
+  rawTokens = regexSplit(rawTokens !== undefined ?
+    rawTokens : inputsStr, SPLIT_PATTERN_1, SPLIT_PATTERN_1);
+  // Second pass splits out the last whilespace char or "६".
+  rawTokens  = regexSplit(rawTokens, SPLIT_PATTERN_2, SPLIT_PATTERN_2);
+
+  if (unsplittableTokens && unsplittableTokens.length > 0) {
+    // Replace special tokens alternate with originals.
+    unsplittableTokens.forEach((token, idx) => {
+      const alt = alts[idx];
+      const escapedAlt = alt.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
```

Comment:
It might be worth having a function that does this regex escaping since it's used in two places and not easy to remember.